x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleService.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.snapshotlifecycle;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.LocalNodeMasterListener;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.core.scheduler.CronSchedule;
+import org.elasticsearch.xpack.core.scheduler.SchedulerEngine;
+
+import java.io.Closeable;
+import java.time.Clock;
+import java.util.Map;
+
+/**
+ * {@code SnapshotLifecycleService} manages snapshot policy scheduling and triggering of the
+ * {@link SnapshotLifecycleTask}. It reacts to new policies in the cluster state by scheduling a
+ * task according to the policy's schedule.
+ */
+public class SnapshotLifecycleService implements LocalNodeMasterListener, Closeable, ClusterStateListener {
```

Comment:
It might be good for this to have a test similar to `IndexLifecycleServiceTests#prepareServices` to make sure it attaches the listener, but I'd also be fine leaving that as a TODO until there's something real to test.

Reply:
Oh totally, I am planning on testing all of this, right now though, I think the implementation may change, so I'll leave it as a future TODO